File: backend/app/schemas/payment.py
"""Payment preview request and exact allocation response schemas."""

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.common import to_camel


class PaymentPreviewRequest(BaseModel):
    """Candidate payment values accepted before an immutable confirmation."""

    amount: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    effective_date: date
    installment_id: str | None = Field(default=None, min_length=36, max_length=36)

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    @field_validator("amount", mode="before")
    @classmethod
    def reject_float_amount(cls, value: Any) -> Any:
        """Require exact JSON decimal strings or integers, never binary floats."""
        if isinstance(value, float):
            raise ValueError("must be sent as an exact decimal string")
        return value

    @field_validator("installment_id")
    @classmethod
    def validate_installment_id(cls, value: str | None) -> str | None:
        """Canonicalize an optional target installment UUID."""
        return None if value is None else str(UUID(value))


class PaymentCreate(PaymentPreviewRequest):
    """Immutable payment confirmation with a retry-safe request UUID."""

    request_id: str = Field(min_length=36, max_length=36)
    note: str | None = Field(default=None, max_length=500)

    @field_validator("request_id")
    @classmethod
    def validate_request_id(cls, value: str) -> str:
        return str(UUID(value))


class PaymentReversalCreate(BaseModel):
    """Reasoned, retry-safe request to reverse one complete payment."""

    request_id: str = Field(min_length=36, max_length=36)
    effective_date: date
    reason: str = Field(min_length=3, max_length=500)

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    @field_validator("request_id")
    @classmethod
    def validate_request_id(cls, value: str) -> str:
        return str(UUID(value))

    @field_validator("reason")
    @classmethod
    def normalize_reason(cls, value: str) -> str:
        reason = value.strip()
        if len(reason) < 3:
            raise ValueError("must contain at least 3 non-whitespace characters")
        return reason


class PaymentPreviewResponse(BaseModel):
    """Authoritative, non-persisted allocation shown before confirmation."""

    loan_id: str
    installment_id: str
    payment_amount: Decimal
    effective_date: date
    period_start_date: date
    accrual_start_date: date
    due_date: date
    scheduled_period_days: int
    elapsed_days: int
    days_early: int
    overdue_days: int
    periodic_rate: Decimal
    accrued_interest: Decimal
    carried_interest_before: Decimal
    total_interest_before: Decimal
    principal_before: Decimal
    applied_interest: Decimal
    applied_principal: Decimal
    unapplied_credit: Decimal
    interest_after: Decimal
    principal_after: Decimal
    scheduled_payment: Decimal
    amount_above_scheduled: Decimal
    next_period_interest: Decimal
    is_payoff: bool

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class PaymentAllocationResponse(BaseModel):
    """Persisted interest-first allocation snapshot."""

    interest_before: Decimal
    principal_before: Decimal
    applied_interest: Decimal
    applied_principal: Decimal
    unapplied_credit: Decimal
    interest_after: Decimal
    principal_after: Decimal
    overdue_days: int
    scheduled_period_days: int

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, from_attributes=True)


class PaymentResponse(BaseModel):
    """One immutable payment ledger entry and its allocation."""

    id: str
    request_id: str
    loan_id: str
    installment_id: str | None
    recorded_by_user_id: str
    reversal_of_payment_id: str | None
    entry_type: Literal["Payment", "Reversal"]
    amount: Decimal
    effective_date: date
    note: str | None
    created_at: datetime
    allocation: PaymentAllocationResponse

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, from_attributes=True)


class PaymentReversalResponse(PaymentResponse):
    """A reversal ledger entry linked permanently to its original payment."""

    reversal_of_payment_id: str
    entry_type: Literal["Reversal"]


class PaymentPage(BaseModel):
    """Backward-compatible paginated payment collection envelope."""

    items: list[PaymentResponse]
    total: int
    offset: int
    limit: int

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

File: lib/features/loans/domain/models/payment.dart
/// Authoritative payment allocation returned before confirmation.
class PaymentPreview {
  const PaymentPreview({
    required this.loanId,
    required this.installmentId,
    required this.paymentAmount,
    required this.effectiveDate,
    required this.dueDate,
    required this.daysEarly,
    required this.overdueDays,
    required this.accruedInterest,
    required this.totalInterestBefore,
    required this.principalBefore,
    required this.appliedInterest,
    required this.appliedPrincipal,
    required this.unappliedCredit,
    required this.interestAfter,
    required this.principalAfter,
    required this.amountAboveScheduled,
    required this.nextPeriodInterest,
    required this.isPayoff,
  });

  final String loanId;
  final String installmentId;
  final String paymentAmount;
  final String effectiveDate;
  final String dueDate;
  final int daysEarly;
  final int overdueDays;
  final String accruedInterest;
  final String totalInterestBefore;
  final String principalBefore;
  final String appliedInterest;
  final String appliedPrincipal;
  final String unappliedCredit;
  final String interestAfter;
  final String principalAfter;
  final String amountAboveScheduled;
  final String nextPeriodInterest;
  final bool isPayoff;

  factory PaymentPreview.fromJson(Map<String, dynamic> json) => PaymentPreview(
    loanId: _string(json, 'loanId'),
    installmentId: _string(json, 'installmentId'),
    paymentAmount: _decimal(json, 'paymentAmount'),
    effectiveDate: _string(json, 'effectiveDate'),
    dueDate: _string(json, 'dueDate'),
    daysEarly: _integer(json, 'daysEarly'),
    overdueDays: _integer(json, 'overdueDays'),
    accruedInterest: _decimal(json, 'accruedInterest'),
    totalInterestBefore: _decimal(json, 'totalInterestBefore'),
    principalBefore: _decimal(json, 'principalBefore'),
    appliedInterest: _decimal(json, 'appliedInterest'),
    appliedPrincipal: _decimal(json, 'appliedPrincipal'),
    unappliedCredit: _decimal(json, 'unappliedCredit'),
    interestAfter: _decimal(json, 'interestAfter'),
    principalAfter: _decimal(json, 'principalAfter'),
    amountAboveScheduled: _decimal(json, 'amountAboveScheduled'),
    nextPeriodInterest: _decimal(json, 'nextPeriodInterest'),
    isPayoff: _boolean(json, 'isPayoff'),
  );
}

/// Immutable confirmed payment and its allocation snapshot.
class LoanPayment {
  const LoanPayment({
    required this.id,
    required this.requestId,
    required this.loanId,
    required this.installmentId,
    required this.entryType,
    required this.reversalOfPaymentId,
    required this.amount,
    required this.effectiveDate,
    required this.note,
    required this.createdAt,
    required this.allocation,
  });

  final String id;
  final String requestId;
  final String loanId;
  final String? installmentId;
  final String entryType;
  final String? reversalOfPaymentId;
  final String amount;
  final String effectiveDate;
  final String? note;
  final String createdAt;
  final PaymentAllocation allocation;

  factory LoanPayment.fromJson(Map<String, dynamic> json) {
    final rawAlloc = json['allocation'];
    final allocMap = rawAlloc is Map
        ? Map<String, dynamic>.from(rawAlloc)
        : const <String, dynamic>{};
    return LoanPayment(
      id: _string(json, 'id'),
      requestId: _string(json, 'requestId'),
      loanId: _string(json, 'loanId'),
      installmentId: (json['installmentId'] ?? json['installment_id'])
          ?.toString(),
      entryType: _string(json, 'entryType', fallback: 'Payment'),
      reversalOfPaymentId:
          (json['reversalOfPaymentId'] ?? json['reversal_of_payment_id'])
              ?.toString(),
      amount: _decimal(json, 'amount'),
      effectiveDate: _string(json, 'effectiveDate'),
      note: json['note']?.toString(),
      createdAt: _string(json, 'createdAt'),
      allocation: PaymentAllocation.fromJson(allocMap),
    );
  }
}

class PaymentAllocation {
  const PaymentAllocation({
    required this.appliedInterest,
    required this.appliedPrincipal,
    required this.unappliedCredit,
    required this.interestAfter,
    required this.principalAfter,
    required this.overdueDays,
  });

  final String appliedInterest;
  final String appliedPrincipal;
  final String unappliedCredit;
  final String interestAfter;
  final String principalAfter;
  final int overdueDays;

  factory PaymentAllocation.fromJson(Map<String, dynamic> json) =>
      PaymentAllocation(
        appliedInterest: _decimal(json, 'appliedInterest'),
        appliedPrincipal: _decimal(json, 'appliedPrincipal'),
        unappliedCredit: _decimal(json, 'unappliedCredit'),
        interestAfter: _decimal(json, 'interestAfter'),
        principalAfter: _decimal(json, 'principalAfter'),
        overdueDays: _integer(json, 'overdueDays'),
      );
}

String _string(Map<String, dynamic> json, String key, {String fallback = ''}) {
  final value = json[key] ?? json[_toSnake(key)];
  if (value != null && value.toString().isNotEmpty) return value.toString();
  if (key == 'requestId') return json['id']?.toString() ?? fallback;
  return fallback;
}

String _decimal(
  Map<String, dynamic> json,
  String key, {
  String fallback = '0.00',
}) {
  final value = json[key] ?? json[_toSnake(key)];
  if (value != null) {
    final str = value.toString();
    if (RegExp(r'^-?\d+(?:\.\d+)?$').hasMatch(str)) return str;
    final parsed = double.tryParse(str);
    if (parsed != null) return parsed.toStringAsFixed(2);
  }
  return fallback;
}

int _integer(Map<String, dynamic> json, String key, {int fallback = 0}) {
  final value = json[key] ?? json[_toSnake(key)];
  if (value is num) return value.toInt();
  if (value is String) return int.tryParse(value) ?? fallback;
  return fallback;
}

bool _boolean(Map<String, dynamic> json, String key, {bool fallback = false}) {
  final value = json[key] ?? json[_toSnake(key)];
  if (value is bool) return value;
  if (value is String) return value.toLowerCase() == 'true';
  return fallback;
}

String _toSnake(String str) {
  return str.replaceAllMapped(
    RegExp(r'[A-Z]'),
    (match) => '_${match.group(0)!.toLowerCase()}',
  );
}
